knowledge-base/upload-preview-image.md

```diff
@@ -0,0 +1,303 @@
+</TelerikUpload>
+
+<div class="image-preview">
+    @foreach (var image in UploadImages)
+    {
+        <img src="@image.ImageSource" class="k-border-2 k-border-solid k-border-primary" />
+    }
+</div>
+
+<style>
+    .image-preview {
+        display: flex;
+        gap: 2em;
+        flex-wrap: wrap;
+        align-items: center;
+    }
+
+        .image-preview img {
+            max-width: 200px;
+            max-height: 200px;
+        }
+</style>
+
+@code {
+    private List<string> ImageFileExtensions { get; set; } = new List<string>() { ".jpg", ".jpeg", ".png", ".gif" };
+
+    private int? MaxImageSize { get; set; } = 8 * 1024 * 1024;
+
+    private List<ImageModel> UploadImages { get; set; } = new();
+
+    private void OnUploadSuccess(UploadSuccessEventArgs args)
+    {
+        var fileInfo = args.Files.First();
+    
+        if (args.Operation == UploadOperationType.Upload)
+        {
+            UploadImages.Add(new ImageModel()
+            {
+                ImageId = fileInfo.Id,
+                ImageSource = args.Request.ResponseText
+            });
+        }
+        else
+        {
+            UploadImages.RemoveAll(x => x.ImageId == fileInfo.Id);
+        }
+    }
+
+    private string ToAbsoluteUrl(string url)
+    {
+        return $"{NavigationManager.BaseUri}{url}";
+    }
+
+    public class ImageModel
+    {
+        public string ImageId { get; set; } = string.Empty;
+        public string ImageSource { get; set; } = string.Empty;
+    }
+}
+````
+````Controller
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+
+namespace TelerikBlazorUpload.Controllers
+{
+    [Route("api/[controller]/[action]")]
+    public class UploadController : ControllerBase
+    {
+        public IWebHostEnvironment HostingEnvironment { get; set; }
+
+        public UploadController(IWebHostEnvironment hostingEnvironment)
+        {
+            HostingEnvironment = hostingEnvironment;
+        }
+
+        [HttpPost]
+        public async Task<string> SaveImage(IFormFile files)
+        {
+            string response = "";
+
+            if (files != null)
+            {
+                try
+                {
+                    // OPTION 1. Return a Base64 string and display the image source as a data URI.
+                    // Not efficient for large files.
+
+                    //var imageBytes = new byte[files.Length];
+                    //await files.OpenReadStream().ReadAsync(imageBytes);
+                    //response = $"data:image/{Path.GetExtension(files.FileName)};base64,{Convert.ToBase64String(imageBytes.ToArray())}";
+
+                    // ===============
+
+                    // OPTION 2. Save the image as a file and return the image URL.
+                    // May not be efficient for large files. You can save another smaller version of the file.
```

Comment:
I recommend moving these notes (or even better, duplicating them) in the step-by-step instructions. They seem pretty important.